### PR TITLE
Fix-CacheId-For-Ui-Config-Data

### PR DESCRIPTION
**Pull request by agorbulin**
Make cacheId unique per store, to separate cache of UI components configs per store

### Description (*)
When we tried to customize widget_recently_viewed UI component per theme 
(we added custom prices for Germany, because their laws require additional prices: per liter/kg/etc)
we faced with bug, when wrong UI component config data were loaded from cache, because cacheId is not unique for the same UI component (not per store).

### Fixed Issues ()
there is no created issue

### Manual testing scenarios (*)
Seems can not be reproduced on default Magento.
We set custom js files for widget_recently_viewed, these settings cached for all stores and widget tries to load custom JS files on other stores, that lead to 404 response (these JS files are not generated for other stores)
Custom code:
**Custom_Theme/Magento_Catalog/ui_component/widget_recently_viewed.xml**
```
<listing xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
    <columns name="widget_columns">
        <column name="status" sortOrder="10">
            <argument name="data" xsi:type="array">
                <item name="config" xsi:type="array">
                    <item name="displayArea" xsi:type="string">details-area</item>
                    <item name="bodyTmpl" xsi:type="string">YourVendor_CustomTheme/product/status</item>
                    <item name="component" xsi:type="string">YourVendor_CustomTheme/js/product/status</item>
                </item>
            </argument>
        </column>
    </columns>
</listing>
```

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @agorbulin. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Ui/Config/Data.php [hunk 1]
```diff
@@ -64,26 +65,37 @@ class Data implements \Magento\Framework\Config\DataInterface
...
+        InterpreterInterface $argumentInterpreter,
+        StoreManagerInterface $storeManager = null
     ) {
         $this->readerFactory = $readerFactory;
         $this->cache = $cache;
         $this->serializer = $serializer;
         $this->componentName = $componentName;
         $this->argumentInterpreter = $argumentInterpreter;
-        $this->cacheId = static::CACHE_ID . '_' . $componentName;
+        $this->storeManager = $storeManager ?:
```
Comment (orlangur) [inline app/code/Magento/Ui/Config/Data.php]: No need to assign property which is only used in constructor.

File: app/code/Magento/Ui/Config/Data.php [hunk 2]
```diff
@@ -64,26 +65,37 @@ class Data implements \Magento\Framework\Config\DataInterface
...
+            \Magento\Framework\App\ObjectManager::getInstance()->get(StoreManagerInterface::class);
+        $this->cacheId = static::CACHE_ID . '_' . $componentName . '_' . $this->storeManager->getStore()->getCode();
```
Comment (orlangur) [inline app/code/Magento/Ui/Config/Data.php]: Please use store_id instead, with code there could be a collision and also it is usually longer.

**Review by orlangur — CHANGES_REQUESTED** [review]
> Can be reproduced only during customization of Magento UI components
in custom theme/module

Please provide a minimalistic code snippet which reveals issue so that it can be verified.

File: app/code/Magento/Ui/Config/Data.php (hunk 1 above)
Comment (agorbulin) [reply app/code/Magento/Ui/Config/Data.php]: We set custom js files for widget_recently_viewed, these settings cached for all stores and widget tries to load custom JS files on other stores, that lead to 404 response (these JS files  are not generated for other stores)
Custom code:
**Custom_Theme/Magento_Catalog/ui_component/widget_recently_viewed.xml**
```
<listing xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="urn:magento:module:Magento_Ui:etc/ui_configuration.xsd">
    <columns name="widget_columns">
        <column name="status" sortOrder="10">
            <argument name="data" xsi:type="array">
                <item name="config" xsi:type="array">
                    <item name="displayArea" xsi:type="string">details-area</item>
                    <item name="bodyTmpl" xsi:type="string">YourVendor_CustomTheme/product/status</item>
                    <item name="component" xsi:type="string">YourVendor_CustomTheme/js/product/status</item>
                </item>
            </argument>
        </column>
    </columns>
</listing>
```

File: app/code/Magento/Ui/Config/Data.php (hunk 2 above)
Comment (agorbulin) [reply app/code/Magento/Ui/Config/Data.php]: Done

File: app/code/Magento/Ui/Config/Data.php (hunk 1 above)
Comment (orlangur) [reply app/code/Magento/Ui/Config/Data.php]: @agorbulin please put this to "manual testing scenario" section.

File: app/code/Magento/Ui/Config/Data.php [hunk 3]
```diff
@@ -83,7 +84,8 @@ public function __construct(
         $this->serializer = $serializer;
         $this->componentName = $componentName;
         $this->argumentInterpreter = $argumentInterpreter;
-        $this->cacheId = static::CACHE_ID . '_' . $componentName;
+        $storeManager = \Magento\Framework\App\ObjectManager::getInstance()->get(StoreManagerInterface::class);
```
Comment (orlangur) [inline app/code/Magento/Ui/Config/Data.php]: Still must be passed via constructor. Just don't assign to property.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Ui/Config/Data.php (hunk 3 above)
Comment (agorbulin) [reply app/code/Magento/Ui/Config/Data.php]: Fixed, please check.

File: app/code/Magento/Ui/Config/Data.php (hunk 1 above)
Comment (agorbulin) [reply app/code/Magento/Ui/Config/Data.php]: Done

**Review by orlangur — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @orlangur, thank you for the review.
[ENGCOM-3429](https://jira.corp.magento.com/browse/ENGCOM-3429) has been created to process this Pull Request

**Comment by sivaschenko** [comment]
The pull request introduces performance degradation:
`Open new category page (Admin Category Management): 81.1% (mainline = 51229 | team = 92797) -> DEGRADATION`

**Comment by orlangur** [comment]
@sivaschenko are we using multi-store in performance tests? Probably cache warmup needs to be adjusted to avoid massive cache misses.

**Comment by sivaschenko** [comment]
@agorbulin @orlangur can you please clarify the manual testing scenario. How can I specify UI configuration for a specific store?

**Comment by orlangur** [comment]
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened.
Thanks for collaboration!

**Pull request closed without merging**